crates/noirc_frontend/src/hir/type_check/expr.rs

```diff
@@ -479,59 +485,84 @@ pub fn infix_operand_type_rules(
     errors: &mut Vec<TypeCheckError>,
-) -> Result<Type, String> {
+) -> Result<Type, TypeCheckError> {
+    let make_error = move |msg| TypeCheckError::Unstructured { msg, span };
+
     if op.kind.is_comparator() {
-        return comparator_operand_type_rules(lhs_type, rhs_type, op, errors);
+        return comparator_operand_type_rules(lhs_type, rhs_type, op, errors).map_err(make_error);
     }
 
     use Type::*;
     match (lhs_type, rhs_type)  {
         (Integer(comptime_x, sign_x, bit_width_x), Integer(comptime_y, sign_y, bit_width_y)) => {
             if sign_x != sign_y {
-                return Err(format!("Integers must have the same signedness LHS is {sign_x:?}, RHS is {sign_y:?} "))
+                return Err(make_error(format!("Integers must have the same signedness LHS is {sign_x:?}, RHS is {sign_y:?} ")))
             }
             if bit_width_x != bit_width_y {
-                return Err(format!("Integers must have the same bit width LHS is {bit_width_x}, RHS is {bit_width_y} "))
+                return Err(make_error(format!("Integers must have the same bit width LHS is {bit_width_x}, RHS is {bit_width_y} ")))
             }
             let comptime = comptime_x.and(comptime_y, op.location.span);
             Ok(Integer(comptime, *sign_x, *bit_width_x))
         }
         (Integer(..), FieldElement(..)) | (FieldElement(..), Integer(..)) => {
-            Err("Cannot use an integer and a Field in a binary operation, try converting the Field into an integer".to_string())
+            Err(make_error("Cannot use an integer and a Field in a binary operation, try converting the Field into an integer".to_string()))
         }
         (PolymorphicInteger(comptime, int), other)
         | (other, PolymorphicInteger(comptime, int)) => {
             if let TypeBinding::Bound(binding) = &*int.borrow() {
-                return infix_operand_type_rules(binding, op, other, errors);
+                return infix_operand_type_rules(binding, op, other, span, interner, errors);
+            }
+
+            if op.is_bitwise() && (other.is_bindable() || other.is_field()) {
+                let other = other.follow_bindings();
+
+                // This will be an error if these types later resolve to a Field, or stay
+                // polymorphic as the bit size will be unknown. Delay this error until the function
+                // finishes resolving so we can still allow cases like `let x: u8 = 1 << 2;`.
+                interner.push_delayed_type_check(Box::new(move || {
+                    if other.is_field() {
+                        Err(make_error("Bitwise operations are invalid on Field types. Try casting the operands to a sized integer type first.".into()))
+                    } else if other.is_bindable() {
+                        Err(make_error("The number of bits to use for this bitwise operation is ambiguous. The type of at least one operand should be specified before this point".into()))
```

Comment:
This is kind of a nit pick but technically couldn't we say "The type of at least one operand or the resulting value should be specified before this point". In your example if I changed `let _c = 2 << 4;` to `let _c: u8 = 2 << 4;` this would be valid. Technically the message is not incorrect because casting either operand will also make the statement valid, just was thinking we could make it more descriptive.

Reply:
Yes, good point. The message was originally made before I added the delayed system when it was not previously influenced by the return type, hence the wording. I added the delay system when I ran the tests and saw that our stdlib would need to be changed as it used `1 << 4` or similar quite a bit. Code with literals like that is also common, so I'll reword the error message.